lib/Cryptol.cry

```diff
@@ -323,3 +324,87 @@ groupBy : {each,parts,elem} (fin each) =>
   [parts * each] elem -> [parts][each]elem
 groupBy = split`{parts=parts}
 
+/**
+ * Logical implication
+ */
+(==>) : {a} a -> a -> a
+a ==> b = not a || b
+
+/**
+ * Logical negation
+ */
+not : {a} a -> a
+not a = ~ a
+
+/**
+ * Conjunction
+ */
+and : {n} (fin n) => [n] -> Bit
+and xs = ~zero == xs
+
+/**
+ * Disjuntion
+ */
+or : {n} (fin n) => [n] -> Bit
+or xs = zero != xs
+
```

Comment:
I disagree with specialization at least for not, and, and or.  These function are rather useful in the polymorphic form:

```
Cryptol> not True
False
Cryptol> not [True, True]
0x0
Cryptol> not [True, False]
0x1
```

We couldn't use `not : Bit -> Bit` if we say `not : [a]Bit -> [a]Bit` and that would almost certainly frustrate people.

On the other hand, specialization for implication seems to make sense.  If we want a vector form them we could add `zipWith` and `zipWIth (==>)`, though then I'd want to work out an under-the-covers optimization.